lib/Runtime/Library/SharedArrayBuffer.cpp

```diff
@@ -0,0 +1,577 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+#include "RuntimeLibraryPch.h"
+
+namespace Js
+{
+#if DBG
+    void SharedContents::AddAgent(DWORD_PTR agent)
+    {
+        if (allowedAgents == nullptr)
+        {
+            allowedAgents = HeapNew(SharableAgents, &HeapAllocator::Instance);
+        }
+
+        allowedAgents->Add(agent);
+    }
+
+    bool SharedContents::IsValidAgent(DWORD_PTR agent)
+    {
+        return allowedAgents != nullptr && allowedAgents->Contains(agent);
+    }
+#endif
+
+    void SharedContents::Cleanup()
+    {
+        Assert(refCount == 0);
+        buffer = nullptr;
+        bufferLength = 0;
+#if DBG
+        if (allowedAgents != nullptr)
+        {
+            HeapDelete(allowedAgents);
+            allowedAgents = nullptr;
+        }
+#endif
+
+        if (indexToWaiterList != nullptr)
+        {
+            indexToWaiterList->Map([](uint index, WaiterList *waiters) {
+                if (waiters != nullptr)
+                {
+                    waiters->Cleanup();
+                    HeapDelete(waiters);
+                    waiters = nullptr;
+                }
+            });
+
+            HeapDelete(indexToWaiterList);
+            indexToWaiterList = nullptr;
+        }
+    }
+
+    uint32 SharedArrayBuffer::GetByteLengthFromVar(ScriptContext* scriptContext, Var length)
+    {
+        if (TaggedInt::Is(length))
+        {
+            int32 byteCount = TaggedInt::ToInt32(length);
+            if (byteCount < 0)
+            {
+                JavascriptError::ThrowRangeError(
+                    scriptContext, JSERR_ArrayLengthConstructIncorrect);
+            }
+            return byteCount;
+        }
+
+        return JavascriptConversion::ToUInt32(length, scriptContext);
+    }
+
+    Var SharedArrayBuffer::NewInstance(RecyclableObject* function, CallInfo callInfo, ...)
```

Comment:
There are few subtle differences in these functions. I feel like it is best to keep them separate as the spec could evolve and it would be easier to make those changes. Majority of this functions are input validation. I could make them a common macro. 

---
In reply to: [77910245](https://github.com/Microsoft/ChakraCore/pull/1533#discussion_r77910245) [](ancestors = 77910245)